File: icamera/src/main/java/me/shouheng/icamera/config/size/AspectRatio.java
package me.shouheng.icamera.config.size;

import android.support.annotation.IntRange;
import android.support.annotation.NonNull;

import me.shouheng.icamera.util.XLog;

/**
 * Aspect ratio
 *
 * @author WngShhng (shouheng2015@gmail.com)
 * @version 2019/4/13 23:08
 */
public class AspectRatio {

    private static final String TAG = "AspectRatio";

    private double ratio;

    public final int widthRatio;

    public final int heightRatio;

    public static AspectRatio of(@NonNull Size size) {
        return new AspectRatio(size.width, size.height);
    }

    public static AspectRatio of(@IntRange(from = 1) int widthRatio, @IntRange(from = 0) int heightRatio) {
        return new AspectRatio(widthRatio, heightRatio);
    }

    public static AspectRatio parse(String s) {
        int position = s.indexOf(':');
        if (position == -1) {
            throw new IllegalArgumentException("Malformed aspect ratio: " + s);
        }
        try {
            int x = Integer.parseInt(s.substring(0, position));
            int y = Integer.parseInt(s.substring(position + 1));
            XLog.d(TAG, "Parsed aspect ratio from string as " + x + " : " + y);
            return AspectRatio.of(x, y);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Malformed aspect ratio: " + s, e);
        }
    }

    private AspectRatio(int widthRatio, int heightRatio) {
        this.widthRatio = widthRatio;
        this.heightRatio = heightRatio;
    }

    public double ratio() {
        if (ratio == 0) {
            ratio = (double) heightRatio / widthRatio;
        }
        return ratio;
    }

    public AspectRatio inverse() {
        return AspectRatio.of(heightRatio, widthRatio);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        AspectRatio that = (AspectRatio) o;

        return Double.compare(that.ratio(), ratio()) == 0;
    }

    @Override
    public int hashCode() {
        long temp = Double.doubleToLongBits(ratio());
        return (int) (temp ^ (temp >>> 32));
    }

    @NonNull
    @Override
    public String toString() {
        return "AspectRatio{" +
                "ratio=" + ratio +
                ", widthRatio=" + widthRatio +
                ", heightRatio=" + heightRatio +
                '}';
    }
}

File: icamera/src/main/java/me/shouheng/icamera/config/size/AspectRatio.kt
package me.shouheng.icamera.config.size

import android.support.annotation.IntRange
import me.shouheng.icamera.util.XLog.d

/**
 * Aspect ratio
 *
 * @author WngShhng (shouheng2015@gmail.com)
 * @version 2019/4/13 23:08
 */
class AspectRatio private constructor(val widthRatio: Int, val heightRatio: Int) {

    private var ratio = 0.0

    fun ratio(): Double {
        if (ratio == 0.0) {
            ratio = heightRatio.toDouble() / widthRatio
        }
        return ratio
    }

    fun inverse(): AspectRatio = of(heightRatio, widthRatio)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as AspectRatio
        return java.lang.Double.compare(that.ratio(), ratio()) == 0
    }

    override fun hashCode(): Int {
        val temp = java.lang.Double.doubleToLongBits(ratio())
        return (temp xor (temp ushr 32)).toInt()
    }

    override fun toString(): String {
        return "AspectRatio{" +
                "ratio=" + ratio +
                ", widthRatio=" + widthRatio +
                ", heightRatio=" + heightRatio +
                '}'
    }

    companion object {

        fun of(size: Size): AspectRatio = AspectRatio(size.width, size.height)

        @JvmStatic
        fun of(@IntRange(from = 1) wr: Int, @IntRange(from = 0) hr: Int): AspectRatio {
            return AspectRatio(wr, hr)
        }

        fun parse(s: String): AspectRatio {
            val position = s.indexOf(':')
            require(position != -1) { "Malformed aspect ratio: $s" }
            return try {
                val x = s.substring(0, position).toInt()
                val y = s.substring(position + 1).toInt()
                d("AspectRatio", "Parsed aspect ratio from string as $x : $y")
                of(x, y)
            } catch (e: NumberFormatException) {
                throw IllegalArgumentException("Malformed aspect ratio: $s", e)
            }
        }
    }

}